### ISSUE TITLE
Site Editor - Update `< Dashboard` in wp-admin mode to return to Calypso.

### ISSUE THREAD

**Issue by Addison-Stavlo** [Automattic/wp-calypso#64131]
Currently, we are updating the `< Dashboard` button in the site editor in the calypso iframe.

This means that only editor instances accessed through calypso (`wordpress.com/site-editor/{SiteURL}`) will land at "My Home" after selecting to exit to the Dashboard. While sites that may get redirected to the wp-admin site editor instance will lead users back into wp-admin.

We want to update this so even the wp-admin editor's `< Dashboard` button leads users back to "My Home" in calypso.

![Screen Shot 2022-05-27 at 3 54 33 PM](https://user-images.githubusercontent.com/28742426/170781025-eabcd845-a106-43c9-bf99-5f5eb82b8d23.png)

**Comment by jeyip** [comment]
Created a draft phab diff to address this problem D83905-code

**Comment by agrullon95** [comment]
Findings and questions (@jeyip and I):

Since calypso-related data would not be available in wp-admin, we decided to investigate how the site-editor inside the calypso iframe was retrieving its `closeURL` value. The `calypsoify-iframe.jsx` component is using the `getEditorCloseConfig` selector to retrieve the `closeURL`which was essentially the last non editor URL (The URL of the previous page*). The logic for storing the last non editor URL can be seen in: [last-non-editor-route/reducer.js](https://github.com/Automattic/wp-calypso/blob/9fb7ac149e55361c0c3a9639ad0caa5e15fc88b0/client/state/route/last-non-editor-route/reducer.js#L15)

Further analysis reveals that the lastNonEditor reducer "ROUTE_SET" action case logic will set the `closeURL` value to the "My Home" URL if the user did not enter the site-editor page from `/page(s)`, `/post(s)`, or `/block-editor `(based on regex match for these three paths). This means the user will not be able to properly navigate back to the previous page for any other page (like the /themes). This could be intentional, but we have a feeling it is a bug that may have not had a major impact since the introduction of the `/site-editor` route.

In conclusion, we cannot reuse the closeURL logic to fix the wp-admin dashboard issue due to its current implementation (and potential bug).

Possible solution:

What do you think about using the hard-coded "My home" URL value for the Dashboard button that you suggested earlier?

If we **always** want to redirect the user to "My Home", then this would be easier to implement and would not rely on the changes made to lastNonEditor reducer (Which may not fix the dashboard button issue).

**Comment by jeyip** [comment]
> Further analysis reveals that the lastNonEditor reducer "ROUTE_SET" action case logic will set the closeURL value to the "My Home" URL if the user did not enter the site-editor page from /page(s), /post(s), or /block-editor (based on regex match for these three paths). This means the user will not be able to properly navigate back to the previous page for any other page (like the /themes). This could be intentional, but we have a feeling it is a bug that may have not had a major impact since the introduction of the /site-editor route.

To clarify, here's a draft PR describing and fixing the bug we were seeing. https://github.com/Automattic/wp-calypso/pull/65577

> In conclusion, we cannot reuse the closeURL logic to fix the wp-admin dashboard issue due to its current implementation (and potential bug).

To clarify this as well, we _might_ be able to reuse the `getEditorCloseConfig` logic somehow, but the emphasis here is the fact that the extra logic provided by `getEditorCloseConfig` wasn't even working for the /site-editor since, well, the entire time the /site-editor route has existed. No one seems to have noticed the different behavior between the post and the site editor, and it seems like a small quality of life improvement, so is it actually worth the effort to reuse `getEditorCloseConfig` in wp-admin? Our intuition is that it probably isn't.

**Comment by Addison-Stavlo** [comment]
Interesting findings here! Thanks @agrullon95 and @jeyip !

The most important thing to do here is to ensure that the site editor from wp-admin goes back to calypso interfaces in some way.  Right now if the `/site-editor/$Site_URL` route redirects the user to the `$site_URL/wp-admin/site-editor.php?` version (which could happen for various reasons), the `< Dashboard` button leads them back into wp-admin instead of calypso.

Fixing the behavior of `getEditorCloseConfig` is nice, but as you said no one has really noticed it and its a small QOL improvement. The fix is simple though so we might as well go with it.

Having the same closeUrl logic in wp-admin would be nice to have, but really the important thing is it goes back to calypso. The wp-admin version always leading back to "My Home" would be fine for what we are trying to resolve here.

**Comment by jeyip** [comment]
Fix is deployed and working as expected on simple sites.

Not working on the atomic site I tested yet, but I know that `wpcom-block-editor` might be served through a cache that needs to be invalidated over time. p9o2xV-1G6-p2 Will check in periodically ( probably in 24h )

**Comment by jeyip** [comment]
Working on atomic 👍
